cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,414 @@ struct ReplaceWithMaskFunctor {
+    switch (batch[0].kind()) {
+      case Datum::ARRAY: {
+        auto array_input = *batch[0].array();
+        int64_t last_valid_value_offset = -1;
+        return FillNullBackwardArray(ctx, array_input, out, array_input,
+                                     &last_valid_value_offset);
+      }
+      case Datum::CHUNKED_ARRAY: {
+        return FillNullBackwardChunkedArray(ctx, batch[0].chunked_array(), out);
+      }
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for drop_null operation: "
+        "values=",
+        batch[0].ToString());
+  }
+
+  static Status FillNullBackwardArray(KernelContext* ctx, ArrayData& array, Datum* out,
+                                      const ArrayData& last_valid_value_chunk,
+                                      int64_t* last_valid_value_offset) {
+    ArrayData* output = out->array().get();
+    output->length = array.length;
+    int8_t direction = -1;
+
+    if (array.MayHaveNulls()) {
+      ARROW_ASSIGN_OR_RAISE(
+          auto reversed_bitmap,
+          arrow::internal::ReverseBitmap(ctx->memory_pool(), array.buffers[0]->data(),
+                                         array.offset, array.length));
+      return FillNullExecutor<Type>::ExecFillNull(
+          ctx, array, reversed_bitmap->data(), output, direction, last_valid_value_chunk,
+          last_valid_value_offset);
+    } else {
+      if (array.length > 0) {
+        *last_valid_value_offset = direction != 1 ? 0 : array.length - 1;
+      }
+      *output = array;
+    }
+    return Status::OK();
+  }
+
+  static Status FillNullBackwardChunkedArray(KernelContext* ctx,
+                                             const std::shared_ptr<ChunkedArray>& values,
+                                             Datum* out) {
+    if (values->null_count() == 0) {
+      *out = Datum(values);
+      return Status::OK();
+    }
+    if (values->null_count() == values->length()) {
+      *out = Datum(values);
+      return Status::OK();
+    }
+    std::vector<std::shared_ptr<Array>> new_chunks;
+
+    if (values->length() > 0) {
+      auto chunks_length = values->chunks().size();
+      ArrayData array_with_current =
+          *values->chunk(/*first_chunk=*/chunks_length - 1)->data();
+      int64_t last_valid_value_offset = -1;
+      auto chunks = values->chunks();
+      for (int i = chunks_length - 1; i >= 0; --i) {
+        const auto& chunk = chunks[i];
+        auto buffer_size = chunk->length() * bit_width(values->type()->id()) / 8;
+
+        std::unique_ptr<ArrayBuilder> builder;
+        RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), values->type(), &builder));
+        RETURN_NOT_OK(builder->Reserve(chunk->length()));
+        ARROW_ASSIGN_OR_RAISE(auto array_output, builder->Finish());
+        ARROW_ASSIGN_OR_RAISE(array_output->data()->buffers[1],
+                              ctx->Allocate(buffer_size));
```

Comment:
Why use a builder when we already calculated a buffer size?

Reply:
Also, why are we doing this for all types when we know binary-like types will allocate their own buffer?

Reply:
Please note, you can directly create an ArrayData. You don't have to use a builder.